crates/iceberg/src/spec/transform.rs

```diff
@@ -261,6 +269,323 @@ impl Transform {
+    }
+
+    /// Projects a binary expression to a predicate with an adjusted boundary.
+    ///
+    /// Checks if the literal within the given binary expression is
+    /// transformable. If transformable, it proceeds to potentially adjust
+    /// the boundary of the expression based on the comparison operator (`op`).
+    /// The potential adjustments involve incrementing or decrementing the
+    /// literal value and changing the `PredicateOperator` itself to its
+    /// inclusive variant.
+    fn project_binary_with_adjusted_boundary<T>(
+        &self,
+        name: String,
+        expr: &BinaryExpression<T>,
+        func: &BoxedTransformFunction,
+        width: Option<u32>,
+    ) -> Result<Option<Predicate>> {
+        if !self.can_transform(expr.literal()) {
+            return Ok(None);
+        }
+
+        let op = &expr.op();
+        let datum = &expr.literal();
+
+        if let Some(boundary) = Self::adjust_boundary(op, datum)? {
+            let transformed_projection = func.transform_literal_result(&boundary)?;
+
+            let adjusted_projection =
+                self.adjust_time_projection(op, datum, &transformed_projection);
+
+            let adjusted_operator = Self::adjust_operator(op, datum, width);
+
+            if let Some(op) = adjusted_operator {
+                let predicate = match adjusted_projection {
+                    None => Predicate::Binary(BinaryExpression::new(
+                        op,
+                        Reference::new(name),
+                        transformed_projection,
+                    )),
+                    Some(AdjustedProjection::Single(d)) => {
+                        Predicate::Binary(BinaryExpression::new(op, Reference::new(name), d))
+                    }
+                    Some(AdjustedProjection::Set(d)) => Predicate::Set(SetExpression::new(
+                        PredicateOperator::In,
+                        Reference::new(name),
+                        d,
+                    )),
+                };
+                return Ok(Some(predicate));
+            }
+        };
+
+        Ok(None)
+    }
+
+    /// Projects a set expression to a predicate,
+    /// applying a transformation to each literal in the set.
+    fn project_set<T>(
+        &self,
+        expr: &SetExpression<T>,
+        name: String,
+        func: &BoxedTransformFunction,
+    ) -> Result<Option<Predicate>> {
+        if expr.op() != PredicateOperator::In
+            || expr.literals().iter().any(|d| !self.can_transform(d))
+        {
+            return Ok(None);
+        }
+
+        let mut new_set = FnvHashSet::default();
+
+        for lit in expr.literals() {
+            let datum = func.transform_literal_result(lit)?;
+
+            if let Some(AdjustedProjection::Single(d)) =
+                self.adjust_time_projection(&PredicateOperator::In, lit, &datum)
+            {
+                new_set.insert(d);
+            };
+
+            new_set.insert(datum);
```

Comment:
Should this be in else branch?

Reply:
@liurenjie1024 
No. The original datum should always be inserted into the `new_set` alongside the adjusted_datum if we have any.
In Java it is done [here](https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/transforms/ProjectionUtil.java#L316-L319).